### ISSUE TITLE
Epochs vs Epsilon Convergence

### ISSUE THREAD

**Issue by dsrishti** [opendp/smartnoise-sdk#477]
Hello Team,

Just wanted to check with you in regards to synthesizing datasets using snsynth, specifically CTGAN based synthesizers. We want to know if there is a way for us to control the convergence rate to higher epsilons within given number of epochs. 

Epoch 1998, Loss G: 0.6126, Loss D: 1.4129
epsilon is 73.71666310336865, alpha is 1.4
Epoch 1999, Loss G: 0.6118, Loss D: 1.4134
epsilon is 73.74491499271095, alpha is 1.4
Epoch 2000, Loss G: 0.6107, Loss D: 1.4092
epsilon is 73.77316688205325, alpha is 1.4

When we gave an epsilon of 100(as input), due to the limitation on # of epochs, it only reached to epsilon of 73.77.  Is there a way for us to control the # of epochs and still obtain a target epsilon value of 100? Because, if we can control this, it will reduce our processing times significantly!

**Comment by joshua-oss** [comment]
Tagging @AprilXiaoyanLiu

**Comment by AprilXiaoyanLiu** [comment]
> Hello Team,
> 
> Just wanted to check with you in regards to synthesizing datasets using snsynth, specifically CTGAN based synthesizers. We want to know if there is a way for us to control the convergence rate to higher epsilons within given number of epochs.
> 
> Epoch 1998, Loss G: 0.6126, Loss D: 1.4129 epsilon is 73.71666310336865, alpha is 1.4 Epoch 1999, Loss G: 0.6118, Loss D: 1.4134 epsilon is 73.74491499271095, alpha is 1.4 Epoch 2000, Loss G: 0.6107, Loss D: 1.4092 epsilon is 73.77316688205325, alpha is 1.4
> 
> When we gave an epsilon of 100(as input), due to the limitation on # of epochs, it only reached to epsilon of 73.77. Is there a way for us to control the # of epochs and still obtain a target epsilon value of 100? Because, if we can control this, it will reduce our processing times significantly!

Hello @dsrishti ,  do you mind letting me know which loss function you are currently using? cross entropy loss sometimes converged faster with relatively similar performance as Wasserstein loss. I think you probably can tune your privacy hyperparameters for example, noise_multipliers, max_grad_norm to adjust the speed of budget spent.

**Comment by dsrishti** [comment]
Hi @AprilXiaoyanLiu ,

We did use 'cross_entropy' loss for convergence. So, within the syntax of DPCTGAN,

    def __init__(
        self,
        embedding_dim=128,
        gen_dim=(256, 256),
        dis_dim=(256, 256),
        l2scale=1e-6,
        batch_size=500,
        epochs=300,
        pack=1,
        log_frequency=True,
        disabled_dp=False,
        target_delta=None,
        **_sigma=5,
        max_per_sample_grad_norm=1.0,_**
        epsilon=1,
        verbose=True,
        loss="cross_entropy",
    )
sigma and max_per_sample_grad_norm should allow us to control the speed of epsilon spent. 

Thank you, will test it out and reach out if we face any issues!!
